src/main/scala/io/iohk/ethereum/network/PeerManagerActor.scala

```diff
@@ -69,14 +90,50 @@ class PeerManagerActor(
     peer
   }
 
+  def getPeers(): Future[Peers] = {
+    implicit val timeout = Timeout(2.seconds)
+
+    Future.traverse(peers.values) { peer =>
+      (peer.ref ? PeerActor.GetStatus)
+        .mapTo[PeerActor.StatusResponse]
+        .map(sr => (peer, sr.status))
+    }.map(r => Peers.apply(r.toMap))
+  }
+
+  def tryDiscardPeersToFreeUpLimit(): Future[Unit] = {
+    getPeers() flatMap { peers =>
+      val peersToPotentiallyDisconnect = peers.peers.filter {
+        case (_, Handshaking(numRetries)) if numRetries > 0 => true /* still handshaking and retried at least once */
+        case (_, PeerActor.Status.Disconnected) => true /* already disconnected */
+        case _ => false
+      }
+
+      if (peers.peers.size - peersToPotentiallyDisconnect.size >= peerConfiguration.maxPeers) {
+        Future.failed(new RuntimeException("Too many peers"))
```

Comment:
Sounds a good strategy, thanks for the explanation :+1:
Regarding this, while testing this branch I found out that we were connected to 6 peers while `max-peers=3` with the status of 5 of them being `Connecting` and `Handshaked` for the remaining peer. Shouldn't that be avoided if I understood correctly?